File: day9/main.cpp
#include <iostream>
#include <cstdint>
#include <string>
#include <vector>

uint64_t get_file_compactness(uint32_t file_id, uint32_t file_size_in_blocks, uint32_t blk_offset) {
    uint64_t res = 0;
    while (file_size_in_blocks > 0) {
        res += blk_offset * file_id;
        blk_offset++;
        file_size_in_blocks--;
    }
    return res;
}

uint64_t get_disk_compactness_b(std::string disk_map) {
    uint64_t res = 0;
    uint32_t i = 0;
    uint32_t blk_offset = 0;
    uint32_t j = disk_map.size() - 1;
    std::string disk_map_copy = disk_map;
    std::vector<uint32_t> blk_offsets(disk_map.size(), 0);
    for (uint32_t i = 1; i < disk_map.size(); i++) {
        blk_offsets[i] = blk_offsets[i - 1] + (static_cast<uint32_t>(disk_map[i - 1]) - '0');
    }
    while (j > 0) {
        uint32_t file_size_in_blocks = static_cast<uint32_t>(disk_map_copy[j]) - '0';
        uint32_t file_id = j / 2;
        for (uint32_t i = 1; i < j; i += 2) {
            uint32_t free_space_in_blocks = static_cast<uint32_t>(disk_map_copy[i]) - '0';
            if (free_space_in_blocks >= file_size_in_blocks) {
                res += get_file_compactness(file_id, file_size_in_blocks, blk_offsets[i]);
                blk_offsets[i] += file_size_in_blocks;
                disk_map_copy[i] = static_cast<char>('0' + (free_space_in_blocks - file_size_in_blocks));
                disk_map_copy[j] = '0';
                break;
            }
        }
        j -= 2;
    }
    for (uint32_t i = 0; i < disk_map.size(); i += 2) {
        uint32_t file_size_in_blocks = static_cast<uint32_t>(disk_map_copy[i]) - '0';
        uint32_t file_id = i / 2;
        res += get_file_compactness(file_id, file_size_in_blocks, blk_offsets[i]);
    }
    return res;
}

uint64_t get_disk_compactness_a(std::string disk_map) {
    uint64_t res = 0;
    uint32_t i = 0;
    uint32_t j = disk_map.size() - 1;
    uint32_t blk_offset = 0;
    while (i <= j) {
        if (i % 2 == 0) {
            uint32_t file_size_in_blocks = static_cast<uint32_t>(disk_map[i]) - '0';
            uint32_t file_id = i / 2;
            res += get_file_compactness(file_id, file_size_in_blocks, blk_offset);
            blk_offset += file_size_in_blocks;
            ++i;
        } else {
            uint32_t file_size_in_blocks = static_cast<uint32_t>(disk_map[j]) - '0';
            uint32_t file_id = j / 2;
            uint32_t free_space_in_blocks = static_cast<uint32_t>(disk_map[i]) - '0';
            uint32_t num_blocks = std::min(file_size_in_blocks, free_space_in_blocks);
            res += get_file_compactness(file_id, num_blocks, blk_offset);
            blk_offset += num_blocks;
            if (free_space_in_blocks == file_size_in_blocks) {
                j -= 2;
                ++i;
            } else if (free_space_in_blocks < file_size_in_blocks) {
                i++;
                disk_map[j] = static_cast<char>('0' + (file_size_in_blocks - num_blocks));
            } else {
                j -= 2;
                disk_map[i] = static_cast<char>('0' + (free_space_in_blocks - num_blocks));
            }
        }
    }

    return res;
}

int main() {
    std::string disk_map;
    std::cin >> disk_map;

    uint64_t res_a = get_disk_compactness_a(disk_map);
    uint64_t res_b = get_disk_compactness_b(disk_map);

    std::cout << res_a << '\n';
    std::cout << res_b << '\n';
    return 0;
}

File: day9/main.rs
use std::io;

fn get_file_compactness(file_id: u32, file_size_in_blocks: u32, blk_offset: u32) -> u64 {
    // Sum of (blk_offset + k) * file_id for k = 0..file_size_in_blocks
    // = file_id * sum(blk_offset + k) for k = 0..n-1
    // = file_id * (n * blk_offset + n*(n-1)/2)
    if file_size_in_blocks == 0 {
        return 0;
    }
    let n = file_size_in_blocks as u64;
    let offset = blk_offset as u64;
    let id = file_id as u64;
    id * (n * offset + n * (n - 1) / 2)
}

fn get_disk_compactness_b(disk_map: &str) -> u64 {
    let chars: Vec<char> = disk_map.chars().collect();
    let mut res = 0u64;
    let mut j = chars.len() - 1;
    let mut chars_copy = chars.clone();
    let mut blk_offsets = vec![0; chars.len()];
    for i in 1..chars.len() {
        blk_offsets[i] = blk_offsets[i - 1] + (chars[i - 1].to_digit(10).unwrap() as u32);
    }
    while j > 0 {
        let file_size_in_blocks = chars_copy[j].to_digit(10).unwrap() as u32;
        let file_id = j as u32 / 2;
        for i in 1..j {
            if i % 2 == 0 {
                continue;
            }
            let free_space_in_blocks = chars_copy[i].to_digit(10).unwrap() as u32;
            if free_space_in_blocks >= file_size_in_blocks {
                res += get_file_compactness(file_id, file_size_in_blocks, blk_offsets[i]);
                blk_offsets[i] += file_size_in_blocks;
                chars_copy[i] = (free_space_in_blocks - file_size_in_blocks).to_string().chars().next().unwrap();
                chars_copy[j] = '0';
                break;
            }
        }
        j -= 2;
    }
    for i in 0..chars_copy.len() {
        if i % 2 == 0 {
            let file_size_in_blocks = chars_copy[i].to_digit(10).unwrap() as u32;
            let file_id = i as u32 / 2;
            res += get_file_compactness(file_id, file_size_in_blocks, blk_offsets[i]);
        }
    }
    res
}


fn get_disk_compactness_a(disk_map: &str) -> u64 {
    let chars: Vec<char> = disk_map.chars().collect();
    let mut res = 0u64;
    let mut i = 0;
    let mut j = chars.len() - 1;
    let mut blk_offset = 0;
    let mut remaining_file_blocks = 0u32;
    let mut remaining_free_blocks = 0u32;

    while i < j {
        if i % 2 == 0 {
            let file_size_in_blocks = chars[i].to_digit(10).unwrap();
            let file_id = i as u32 / 2;
            res += get_file_compactness(file_id, file_size_in_blocks, blk_offset);
            blk_offset += file_size_in_blocks;
            i += 1;
        } else {
            if remaining_file_blocks == 0 {
                remaining_file_blocks = chars[j].to_digit(10).unwrap();
            }
            
            if remaining_free_blocks == 0 {
                remaining_free_blocks = chars[i].to_digit(10).unwrap();
            }
            
            let file_id = j as u32 / 2;
            let num_blocks = std::cmp::min(remaining_file_blocks, remaining_free_blocks);
            res += get_file_compactness(file_id, num_blocks, blk_offset);
            blk_offset += num_blocks;
            
            remaining_file_blocks -= num_blocks;
            remaining_free_blocks -= num_blocks;
            
            if remaining_file_blocks == 0 {
                j -= 2;
            }
            if remaining_free_blocks == 0 {
                i += 1;
            }
        }
    }

    if remaining_file_blocks != 0 {
        let file_id = j as u32 / 2;
        res += get_file_compactness(file_id, remaining_file_blocks, blk_offset);
    }
    res
}

fn main() {
    let mut disk_map = String::new();
    io::stdin().read_line(&mut disk_map).unwrap();
    let disk_map = disk_map.trim();

    let res_a = get_disk_compactness_a(&disk_map);
    let res_b = get_disk_compactness_b(&disk_map);

    println!("{}", res_a);
    println!("{}", res_b);
}